File: 剑指offer-2012年书籍版/Java/No06.java
/**
 * 根据前序遍历和中序遍历建立树
 */

public class No06 {

    public static void main(String[] args) {
        String preOrder = "12473568";
        String midOrder = "47215386";
        BiTree tree = new BiTree(preOrder, midOrder, preOrder.length());
        tree.postRootTraverse(tree.root);
    }

}

class BiTree {
    TreeNode root;

    public BiTree(String preOrder, String midOrder, int count) {
        if (count <= 0) {
            return;
        }
        char c = preOrder.charAt(0);
        int i = 0;
        for (; i < count; i++) {
            if (midOrder.charAt(i) == c)
                break;
        }

        root = new TreeNode(c);
        root.setLchild(new BiTree(preOrder.substring(1, i + 1), midOrder.substring(0, i), i).root);
        root.setRchild(new BiTree(preOrder.substring(i + 1), midOrder.substring(i + 1), count - i - 1).root);
    }


    public void postRootTraverse(TreeNode root) {
        if (root != null) {
            postRootTraverse(root.getLchild());
            postRootTraverse(root.getRchild());
            System.out.print(root.getData());
        }
    }
}

class TreeNode {
    char data;
    TreeNode Lchild;
    TreeNode Rchild;

    public TreeNode(char data) {
        super();
        this.data = data;
    }

    public TreeNode(char data, TreeNode lchild, TreeNode rchild) {
        super();
        this.data = data;
        Lchild = lchild;
        Rchild = rchild;
    }

    public char getData() {
        return data;
    }

    public void setData(char data) {
        this.data = data;
    }

    public TreeNode getLchild() {
        return Lchild;
    }

    public void setLchild(TreeNode lchild) {
        Lchild = lchild;
    }

    public TreeNode getRchild() {
        return Rchild;
    }

    public void setRchild(TreeNode rchild) {
        Rchild = rchild;
    }


}

File: 剑指offer-2012年书籍版/Kotlin/No06.kt
/**
 * 根据前序遍历和中序遍历建立树
 */

class No06 {

    fun main() {
        val preOrder = "12473568"
        val midOrder = "47215386"
        BiTree(preOrder, midOrder).also {
            it.print()
        }
    }
}

class BiTree(preOrder: String, midOrder: String) {
    var root: TreeNode? = null

    init {
        if (preOrder.isEmpty() || midOrder.isEmpty()) {
            root = null
        } else {
            if (preOrder.length == 1) {
                root = TreeNode(preOrder[0])
            } else {
                val rootData = preOrder[0]
                val rootIndex = midOrder.indexOf(rootData)
                root = TreeNode(rootData)
                root?.leftChild = BiTree(preOrder.substring(1, rootIndex + 1), midOrder.substring(0, rootIndex)).root
                root?.rightChild = BiTree(preOrder.substring(rootIndex + 1), midOrder.substring(rootIndex + 1)).root
            }
        }
    }

    fun print() {
        println("前序遍历：========= ")
        printPre(root)
        println("\n中序遍历：========= ")
        printMid(root)
        println("\n后序遍历：========= ")
        printPost(root)
    }

    fun printPre(rootNode: TreeNode?) {
        if (rootNode != null) {
            print(rootNode.data)
            printPre(rootNode.leftChild)
            printPre(rootNode.rightChild)
        }
    }

    fun printMid(rootNode: TreeNode?) {
        if (rootNode != null) {
            printMid(rootNode.leftChild)
            print(rootNode.data)
            printMid(rootNode.rightChild)
        }
    }

    fun printPost(rootNode: TreeNode?) {
        if (rootNode != null) {
            printPost(rootNode.leftChild)
            printPost(rootNode.rightChild)
            print(rootNode.data)
        }
    }
}

class TreeNode(var data: Char = ' ') {
    var leftChild: TreeNode? = null
    var rightChild: TreeNode? = null
}